server/src/services/user.service.ts

```diff
@@ -123,6 +125,35 @@ export class UserService {
     });
   }
 
+  async deleteLicense({ user }: AuthDto): Promise<void> {
+    await this.userRepository.deleteMetadata(user.id, UserMetadataKey.LICENSE);
+  }
+
+  async setLicense(auth: AuthDto, license: LicenseKeyDto): Promise<LicenseResponseDto> {
+    if (!license.licenseKey.startsWith('IMCL-')) {
+      throw new BadRequestException('Invalid license key');
+    }
+    const trimmedLicenseKey = license.licenseKey;
+    const licenseValid = this.cryptoRepository.verifySha256(
+      trimmedLicenseKey,
+      license.activationKey,
+      getClientLicensePublicKey(),
+    );
+
+    if (licenseValid) {
+      await this.userRepository.upsertMetadata(auth.user.id, {
+        key: UserMetadataKey.LICENSE,
+        value: {
+          licenseKey: license.licenseKey,
+          activationKey: license.activationKey,
+          activatedAt: new Date(),
+        },
+      });
+    }
+
+    return { valid: licenseValid };
```

Comment:
Personally I think this should not return a 200 if it doesn't set it. Just throw a bad request, the key is invalid.